[Bug]: Compiler error with jest.config.ts and NodeNext moduleResolution in tsconfig.json

### Version

29.1.2

### Steps to reproduce

1. Create / migrate an existing TS project to `moduleResolution: "NodeNext"` (with `module: "NodeNext")
2. Have the `jest.config.ts` file inside the project
3. Run Jest 

### Expected behavior

Correctly parsed Jest config

### Actual behavior

```
Error: Jest: Failed to parse the TypeScript config file /Users/akwodkiewicz/my-repo/jest.config.ts
  TSError: ⨯ Unable to compile TypeScript:
error TS5110: Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.

    at readConfigFileAndSetRootDir (/Users/akwodkiewicz/my-repo/node_modules/@jest/core/node_modules/jest-config/build/readConfigFileAndSetRootDir.js:136:13)
    at async readConfig (/Users/akwodkiewicz/my-repo/node_modules/@jest/core/node_modules/jest-config/build/index.js:216:18)
    at async readConfigs (/Users/akwodkiewicz/my-repo/node_modules/@jest/core/node_modules/jest-config/build/index.js:404:26)
    at async runCLI (/Users/akwodkiewicz/my-repo/node_modules/@jest/core/build/cli/index.js:182:59)
    at async Object.run (/Users/akwodkiewicz/my-repo/node_modules/jest-cli/build/cli/index.js:155:37)
```

### Additional context

The same error happens for `"moduleResolution": "Node16"` setting:
```
TS5110: Option 'module' must be set to 'Node16' when option 'moduleResolution' is set to 'Node16'.
```

and for `"moduleResolution": "Bundler"`

```
TS5095: Option 'bundler' can only be used when 'module' is set to 'es2015' or later.
```

### Environment

```shell
System:
    OS: macOS 14.1
    CPU: (10) arm64 Apple M1 Pro
  Binaries:
    Node: 18.18.2 - /opt/homebrew/opt/nvm/versions/node/v18.18.2/bin/node
    Yarn: 4.0.2 - /opt/homebrew/opt/nvm/versions/node/v18.18.2/bin/yarn
    npm: 9.8.1 - /opt/homebrew/opt/nvm/versions/node/v18.18.2/bin/npm
    bun: 1.0.7 - ~/.bun/bin/bun
```

I ran into the same issue using `moduleResolution: "bundler"`

```
  TSError: ⨯ Unable to compile TypeScript:
error TS5095: Option 'bundler' can only be used when 'module' is set to 'es2015' or later.
```

I think this is the same issue: https://github.com/jestjs/jest/issues/13350
The issue mentions using `TS_NODE_PROJECT` environment variable to pass in a different tsconfig to set commonjs and node16 for moduleResoluton which also works. https://github.com/jestjs/jest/issues/13350#issuecomment-1737623479

I have not tested your PR #14739 but it makes sense to be using node16 with commonjs and should fix the problem

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -52,6 +52,7 @@
 - `[jest-config]` Support `testTimeout` in project config ([#14697](https://github.com/jestjs/jest/pull/14697))
 - `[jest-config]` Support `coverageReporters` in project config ([#14697](https://github.com/jestjs/jest/pull/14830))
 - `[jest-config]` Allow `reporters` in project config ([#14768](https://github.com/jestjs/jest/pull/14768))
+- `[jest-config]` Allow Node16/NodeNext/Bundler `moduleResolution` in project's tsconfig ([#14739](https://github.com/jestjs/jest/pull/14739))
 - `[jest-each]` Allow `$keypath` templates with `null` or `undefined` values ([#14831](https://github.com/jestjs/jest/pull/14831))
 - `[@jest/expect-utils]` Fix comparison of `DataView` ([#14408](https://github.com/jestjs/jest/pull/14408))
 - `[@jest/expect-utils]` [**BREAKING**] exclude non-enumerable in object matching ([#14670](https://github.com/jestjs/jest/pull/14670))

--- a/e2e/__tests__/typescriptConfigFile.test.ts
+++ b/e2e/__tests__/typescriptConfigFile.test.ts
@@ -8,7 +8,7 @@
 import {tmpdir} from 'os';
 import * as path from 'path';
 import {cleanup, writeFiles} from '../Utils';
-import runJest from '../runJest';
+import runJest, {getConfig} from '../runJest';
 
 const DIR = path.resolve(tmpdir(), 'typescript-config-file');
 
@@ -106,4 +106,20 @@ test('works with multiple typescript configs that import something', () => {
   expect(stderr).toContain('Test Suites: 4 passed, 4 total');
   expect(exitCode).toBe(0);
   expect(stdout).toBe('');
+});
+
+test("works with single typescript config that does not import anything with project's moduleResolution set to Node16", () => {
+  const {configs} = getConfig(
+    'typescript-config/modern-module-resolution',
+    [],
+    {
+      skipPkgJsonCheck: true,
+    },
+  );
+
+  expect(configs).toHaveLength(1);
+  expect(configs[0].displayName).toEqual({
+    color: 'white',
+    name: 'Config from modern ts file',
+  });
 });

--- a/e2e/typescript-config/modern-module-resolution/__tests__/test.js
+++ b/e2e/typescript-config/modern-module-resolution/__tests__/test.js
@@ -1,0 +1,10 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+test('dummy test', () => {
+  expect(1).toBe(1);
+});

--- a/e2e/typescript-config/modern-module-resolution/jest.config.ts
+++ b/e2e/typescript-config/modern-module-resolution/jest.config.ts
@@ -1,0 +1,12 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const config = {
+  displayName: 'Config from modern ts file',
+  testEnvironment: 'node',
+};
+export default config;

--- a/e2e/typescript-config/modern-module-resolution/package.json
+++ b/e2e/typescript-config/modern-module-resolution/package.json
@@ -1,0 +1,3 @@
+{
+  "type": "commonjs"
+}

--- a/e2e/typescript-config/modern-module-resolution/tsconfig.json
+++ b/e2e/typescript-config/modern-module-resolution/tsconfig.json
@@ -1,0 +1,8 @@
+{
+  "compilerOptions": {
+    "target": "ESNext",
+    "module": "Node16",
+    "moduleResolution": "Node16",
+    "strict": true
+  }
+}

--- a/packages/jest-config/src/readConfigFileAndSetRootDir.ts
+++ b/packages/jest-config/src/readConfigFileAndSetRootDir.ts
@@ -119,7 +119,7 @@ async function registerTsNode(): Promise<Service> {
     return tsNode.register({
       compilerOptions: {
         module: 'CommonJS',
-        moduleResolution: 'Node',
+        moduleResolution: 'Node10',
       },
       moduleTypes: {
         '**': 'cjs',

--- a/packages/jest-config/src/readConfigFileAndSetRootDir.ts
+++ b/packages/jest-config/src/readConfigFileAndSetRootDir.ts
@@ -119,6 +119,7 @@ async function registerTsNode(): Promise<Service> {
     return tsNode.register({
       compilerOptions: {
         module: 'CommonJS',
+        moduleResolution: 'Node',
       },
       moduleTypes: {
         '**': 'cjs',

